keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,481 @@
+
+The Metrics and its cache between the EndpointSlice Reconciler and EndpointSlice Mirroring Reconciler are similar but not identical. They will be merged together to provide a re-usable and non-type specific metrics package.
+
+The EndpointSlice controller will be adapted accordingly to support the new way of using the EndpointSlice Reconciler.
+
+Finally, the EndpointSlice Mirroring reconciler will be discarded from `pkg/controller/endpointslicemirroring` and the EndpointSlice Mirroring Controller will use the EndpointSlice reconciler from staging with the appropriate configuration.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Cloud Native Network Function (CNF) vendor, some of my services are handled by custom Service-Proxies/Gateways over secondary networks provided by, for example, [Multus](https://github.com/k8snetworkplumbingwg/multus-cni). IPs configured in the service and registered by the EndpointSlice controller must be only the secondary IPs provided by the secondary network provider (e.g. [Multus](https://github.com/k8snetworkplumbingwg/multus-cni), also see [KEP-3698](https://github.com/kubernetes/enhancements/issues/3698)).
+
+Therefore, it must be possible to disable the default Kubernetes EndpointSlice Controller for certain services and re-use the Kubernetes EndpointSlice reconciler implementation to create a controller for secondary network providers.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The existing behavior will be kept by default, and the Kubernetes EndpointSlice Controller will be disabled only when the service contains this new label. This ensures services without the label to continue to be managed as usual.
+
+This will have no effect on other EndpointSlice controller implementations since they will not be influenced by the presence of this label.
+
+## Design Details
+
+### Well-Known Label
+
+The kube-controller-manager should pass to the Endpoints, EndpointSlice and EndpointSlice Mirroring Controllers an informer selecting services that are not labelled with `service.kubernetes.io/endpoint-controller-name`.
+
+### Flexibility on the EndpointSlice Reconciler Module
+
+New fields will be added to the `Reconciler` struct to cover the needs of the EndpointSlice and EndpointSlice Mirroring Controllers. Unlike the EndpointSlice Mirroring Controller, the EndpointSlice Controller needs placeholders if a service does not contain any endpoint. So a boolean `placeholderEnabled` will be added. The EndpointSlice Mirroring Controller does not verify which resource owns the EndpointSlice before Updating/Deleting it, another boolean `ownershipEnforced` would control this.
```

Comment:
placeholder seems to require a section to explain better the concept https://github.com/kubernetes/enhancements/pull/4771/files#r1685505013

> The EndpointSlice Mirroring Controller does not verify which resource owns the EndpointSlice before Updating/Deleting it, another boolean `ownershipEnforced` would control this.

how do you deal with orphan objects then?

Reply:
The owner reference is related to this function in the reconciler: [staging/src/k8s.io/endpointslice/utils.go#L159](https://github.com/kubernetes/endpointslice/blob/v0.30.3/utils.go#L159). I am not sure why this is required, the EndpointSlice Controller doesn't use this to get the EndpointSlice(s) for a particalar service. It uses only the EndpointSlice labels [pkg/controller/endpointslice/endpointslice_controller.go#L369](https://github.com/kubernetes/kubernetes/blob/v1.30.3/pkg/controller/endpointslice/endpointslice_controller.go#L369). With the ownership required enabled, if an EndpointSlice does not have the owner set, then that EndpointSlice will be deleted (current behavior in the EndpointSlice Reconciler).

Reply:
why should not be always the case? why should it be kept lingering if it is generated from a Service?

Reply:
I agree, this should always be the case. I am not sure why, but the EndpointSlice Mirroring Controller does not require this. The ownership is not enforced, if the EndpointSlice is not owned, the Reconciler will not delete it and create a correct owned one. I made the implementation, and the EndpointSlice Mirroring unit tests are not passing if we enforce the ownership.

Reply:
I see, so is an implementation detail , can you send a PR and set the title to WIP so I can take a look?

Reply:
https://github.com/kubernetes/kubernetes/pull/126304

Sorry if the PR is so large, I was planning to create smaller PRs to propose these changes if this KEP would be accepted.

I moved the unit tests from here: [pkg/controller/endpointslicemirroring/reconciler_test.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslicemirroring/reconciler_test.go) to here: [staging/src/k8s.io/endpointslice/reconciler_test.go#L2274](https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L2274)

If you add the option `WithOwnershipEnforced(true)` for the Reconciler used in the tests ([instanciated here](https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L3484)), then the tests will fail.

To run the tests: `go test -count=1 -run ^TestReconcileWithEndpoints$ k8s.io/endpointslice`

Reply:
@robscott can you please chime in? I'm worried that we do changes without fully understand the implications that bit us later ^^^

Reply:
I get that once a bug has existed for long enough it can be a feature, but it's hard for me to justify any of our controllers ever deleting EndpointSlices they did not claim ownership of. I think a smallish bug fix PR for the EndpointSlice Mirroring controller just to address that would help simplify this PR.

Reply:
I mark this conversation as resolved as the flexibility of the reconciler as been removed from this KEP